miq_cal_skipDays / :skip_days - removal of broken code

himdel opened the pull request:
According to http://docs.dhtmlx.com/api__dhtmlxcalendar_setinsensitivedays.html ,
setInsensitiveDays expects dates, not days of week.

Thus, that functionality as it is always triggers a JS error.

So I'm assuming that we can safely remove it? Or at the very least this shows what needs to get rewritten.

dclarizio commented:
@himdel This functionality is still in use.  I'm not sure if the DHTMLX JS setInsensitiveRange function changed or we changed the code at some point to call the wrong function, but if performance charts are based on a time profile that has a subset of the days of the week, we would need to not allow them to be selected in the calendar.  Perhaps we should change it to use this http://docs.dhtmlx.com/api__dhtmlxcalendar_disabledays.html using the "week" parameter, but we should not remove this feature.  Marking as WIP until corrected.  Thx, Dan

himdel commented:
@dclarizio Right, thanks for the pointer, I didn't find disableDays before.

Ok - changed to fix it and wrap in in a helper... 

 * cal.disableDays is used instead of setInsensitiveDays
   * new js_build_calendar helper handles the javascript for
     miqBuildCalendar
   * every usage of miqBuildCalendar with miq_cal_skipDays is replaced
     by a js_build_calendar call
   * setting skip_days from time_profile_days is handled by another
     helper - converts before different notations for sunday (0 vs 7)
   * created dates don't drop time info, bogus dates are no longer
     allowed
   * miq_cal_* variables are always overwritten, so that the calendar
     can't be tainted by previous state
   * tests are fixed correspondingly

chessbyte commented:
@dclarizio @martinpovolny @h-kataria Please review

chessbyte commented:
@dclarizio @martinpovolny @h-kataria Please review

chessbyte commented:
@dclarizio @martinpovolny @h-kataria Please review

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -2913,6 +2913,13 @@ def patternfly_tab_content(id, active, options = {}, &block)
   end
 
+  def skip_days_from_time_profile(time_profile_days)
+    [1, 2, 3, 4, 5, 6, 7].delete_if do |d|
```
`(1..7).delete_if`...

martinpovolny commented on app/helpers/js_helper.rb:
```diff
@@ -110,4 +110,20 @@ def javascript_unchecked(element)
+    out << "miq_cal_skipDays = #{skip_days};"
+    out << 'miqBuildCalendar();'
+
+    out.join("\n")
```
<<EOD
miq_cal_dateFrom = #{js_format_date(options[:date_from])};
miq_cal_dateTo   = #{js_format_date(options[:date_to])};
miq_cal_skipDays = #{skip_days};
miqBuildCalendar();
EOD

martinpovolny commented:
@himdel : a separate PR for the remaining stuff would be fine.

dclarizio commented:
@himdel Would it be possible to write tests for the `skip_days_from_time_profile` and `js_build_calendar` new helper methods?

himdel commented:
@dclarizio Sure :) .. added tests for the new helpers, and changed the relevant ExplorerPresenter spec to only check that js_build_calendar is called since the specific tests were moved there.

himdel commented:
Note: the test failures (PG::UndefinedColumn) have nothing do to with my changes and come from today's master.

chessbyte commented:
@himdel Restarted the failed Travis jobs

dclarizio commented:
@himdel need to rebase and repush.

himdel commented:
TODO also fix `presenter[:build_calendar]` calls to use a date object, there's at least 2 places using strftime(%Y,%m,%d) meaning they're off by a month.

himdel commented:
marking wip for now, parts were obsoleted by the calendar replacement and there's more issues that should be fixed

@miq-bot add_label wip

dclarizio commented:
@himdel Can you please rebase this?  I have a DB with some time profiles with missing dates that I can give this a try on, so will do once you rebase.  Thx, Dan

himdel commented:
@dclarizio rebased :) I will address the rest in a separate PR, because there are still some places that just set the globals and rely on miqBuildCalendar being called from miqOnLoad. For now, I've renamed miq_cal_skipDays for consistency with the other params and rebased.

himdel commented:
@miq-bot remove_label wip

chessbyte commented:
@himdel @dclarizio What is the status of this PR?

himdel commented:
Ready from my side.. :)

dclarizio commented:
@chessbyte I haven't gotten around to testing this as it involves time profiles with selected days of the week (not all 7). I have a DB, but just haven't had time to get to it . . . it's on my list.

chessbyte commented:
@dclarizio bump

himdel commented:
(hopefully) fixed test failures - js_build_calendar expects a real date object, not `Time.zone.now.to_i * 1000` nor `"1970, 0, 1"`

chessbyte commented:
@himdel Please rebase and push again

himdel commented:
Rebased..

chessbyte commented:
@himdel Please rebase and push again

himdel commented:
@chessbyte rebased... any chance we can merge it soon-ish? This is the 5th rebase..

h-kataria commented:
@himdel, i will try to see if i have environment to test this today.

himdel commented:
Thanks, much appreciated :)

h-kataria commented:
@himdel i am unable to test this PR successfully, i am getting never ending spinner and i see errors in Firebug, these errors might be related to recent toolbars changes @martinpovolny any idea what's causing these errors.

exception caught evaling RJS
	

console.error('exception caught evaling RJS');

jquery_...?body=1 (line 6)
TypeError: ManageIQ.toolbars is null
	

jquery.... > eval (line 2)
script follows:
	

console.error('script follows:')

jquery_...?body=1 (line 8)
ManageIQ.charts.chartData = {"candu":[{"xml":{"data":[[null]],"options":{"title":"No records found for this chart","seriesColors":["#ec7a08","#f4aa00","#006e9c","#0085cf","#3f9c35","#92D400","#ccce00","#ffff00","#925bad","#c0acdc","#350000","#a30000","#cc0000"],"seriesDefaults":{"shadow":false,"rendererOptions":{"dataLabelPositionFactor":0.7,"sliceMargin":2}},"grid":{"drawGridlines":true,"gridLineColor":"#e1e1e1","borderWidth":0,"background":"#fff","shadow":false},"legend":

himdel commented:
@h-kataria , @martinpovolny the new toolbars seem to not use `ManageIQ.toolbars` which was removed .. but..

```
app/controllers/dashboard_controller.rb
580:          page << "ManageIQ.toolbars.center_tb.obj.showItem('#{button_id}');"
585:          page << "ManageIQ.toolbars.center_tb.obj.hideItem('#{button_id}');"
624:      "ManageIQ.toolbars.center_tb.obj.enableItem('#{button_id}'); ManageIQ.toolbars.center_tb.obj.setItemToolTip('#{button_id}', '#{tooltip}');"
627:      "ManageIQ.toolbars.center_tb.obj.disableItem('#{button_id}'); ManageIQ.toolbars.center_tb.obj.setItemToolTip('#{button_id}', '#{tooltip}');"

app/controllers/application_controller/compare.rb
111:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_all');"
112:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_all', false);"
113:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_same');"
114:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_same', false);"
115:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('compare_diff');"
117:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_diff', false);"
118:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_diff', true);"
120:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_all');"
121:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_all', false);"
122:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_diff');"
123:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_diff', false);"
124:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('compare_same');"
126:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_same', false);"
127:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_same', true);"
129:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_diff');"
130:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_diff', false);"
131:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('compare_same');"
132:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_same', false);"
133:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('compare_all');"
135:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_all', false);"
136:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('compare_all', true);"
172:        page << "ManageIQ.toolbars.view_tb.obj.enableItem('compare_expanded');"
173:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_expanded', false);"
174:        page << "ManageIQ.toolbars.view_tb.obj.disableItem('compare_compressed');"
176:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_compressed', false);"
177:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_compressed', true);"
179:        page << "ManageIQ.toolbars.view_tb.obj.enableItem('compare_compressed');"
180:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_compressed', false);"
181:        page << "ManageIQ.toolbars.view_tb.obj.disableItem('compare_expanded');"
183:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_expanded', false);"
184:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('compare_expanded', true);"
200:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('comparemode_details');"
201:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_details', false);"
202:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('comparemode_exists');"
204:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_exists', false);"
205:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_exists', true);"
207:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('comparemode_exists');"
208:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_exists', false);"
209:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('comparemode_details');"
211:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_details', false);"
212:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('comparemode_details', true);"
536:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_all');"
537:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_all', false);"
538:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_same');"
539:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_same', false);"
540:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('drift_diff');"
542:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_diff', false);"
543:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_diff', true);"
545:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_all');"
546:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_all', false);"
547:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_diff');"
548:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_diff', false);"
549:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('drift_same');"
551:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_same', false);"
552:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_same', true);"
554:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_diff');"
555:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_diff', false);"
556:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('drift_same');"
557:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_same', false);"
558:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('drift_all');"
560:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_all', false);"
561:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('drift_all', true);"
625:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('driftmode_details');"
626:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_details', false);"
627:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('driftmode_exists');"
629:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_exists', false);"
630:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_exists', true);"
632:        page << "ManageIQ.toolbars.center_tb.obj.enableItem('driftmode_exists');"
633:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_exists', false);"
634:        page << "ManageIQ.toolbars.center_tb.obj.disableItem('driftmode_details');"
636:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_details', false);"
637:        page << "ManageIQ.toolbars.center_tb.obj.setItemState('driftmode_details', true);"
652:        page << "ManageIQ.toolbars.view_tb.obj.enableItem('drift_expanded');"
653:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_expanded', false);"
654:        page << "ManageIQ.toolbars.view_tb.obj.disableItem('drift_compressed');"
656:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_compressed', false);"
657:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_compressed', true);"
659:        page << "ManageIQ.toolbars.view_tb.obj.enableItem('drift_compressed');"
660:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_compressed', false);"
661:        page << "ManageIQ.toolbars.view_tb.obj.disableItem('drift_expanded');"
663:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_expanded', false);"
664:        page << "ManageIQ.toolbars.view_tb.obj.setItemState('drift_expanded', true);"

app/controllers/application_controller/performance.rb
74:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.showItem('#{pfx}perf_refresh');"
75:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.showItem('#{pfx}perf_reload');"
76:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.enableItem('#{pfx}perf_refresh');"
77:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.enableItem('#{pfx}perf_reload');"
79:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.hideItem('#{pfx}perf_refresh');"
80:          page << "if(ManageIQ.toolbars.center_tb.obj) ManageIQ.toolbars.center_tb.obj.hideItem('#{pfx}perf_reload');"
```

himdel commented:
... oh, already mentioned in #5116

himdel commented:
@h-kataria until that's fixed, could you test by adding `ManageIQ.toolbars = { center_tb: {}, view_tb: {} };` to miq_global.js?

himdel commented:
As for the test failure, I'm quite sure this PR can't be the culprit...

```
  1) Metric as vmware with a small environment and time_profile with Vm hourly performances and Host realtime performances calling perf_rollup to hourly on the Host should rollup Host realtime and Vm hourly into Host hourly rows correctly
     Failure/Error: perf.sys_uptime_absolute_latest.should == 30.0    # pulled from Host realtime
       expected: 30.0
            got: 2.0 (using ==)
     # ./spec/models/metric_spec.rb:718:in `block (7 levels) in <top (required)>'
```

This sounds like the code with the recent deprecations, do you know any more about it?

h-kataria commented:
@chessbyte ready to merge when Travis is greeen.

h-kataria commented:
@chessbyte can you restart vmdb test on this PR.

chessbyte commented:
@h-kataria restarted Travis vmdb suite. Will merge when green
